### PR TITLE
Issue 5: Implement build tool - webpack

### PR DESCRIPTION
**Pull request by jschroed91**
This PR takes on a few items:

- #5 - Webpack
- #16 - Dev server
- #3 - Yarn

This adds several things:

1. yarn as package manager (hence the number of lines in this PR)
2. webpack as build tool
3. express as dev server
4. webpack middleware and hot module reload
5. HtmlWebpackPlugin to inject assets into `index.html`
6. `main.js` as the entry point of the application, which currently just sets the root element's innerHTML

Build Tool Configuration:

`config/project.config.js`

This is where project-specific settings for webpack are configured - it exports a single object with configuration values which are used by the build tools.

This is where global variables, project structure, compiler settings, etc. can all be defined.

`config/environments.config.js`

This file is used to override settings in `config/project.config.js` based on the environment. This is used to set environment-specific settings like the URL of an API endpoint as an example.

`config/webpack.config.js`

This file is where webpack is configured. It exports a single object that is passed directly into webpack.

This is where webpack loaders and plugins can be configured, added, or removed.

Needs documentation:

- [x] Detailed descriptions on the options in `project.config.js`.
- [x] Plugins and loaders set up in the current webpack configuration
- [x] Docs on `bin/compile.js` and `bin/dev-server.js`
- [x] Docs on the webpack entry points and code splitting set up
- [x] Add to README installation and setup docs
- [x] Document how express is used in `server/main.js`
- [x] Document the scripts available in package.json
- [x] Document how `src/main.js` is currently set up

### CODE REVIEW

File: config/webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+// css-loader not to duplicate minimization.
+const BASE_CSS_LOADER = 'css?sourceMap&-minimize';
+
+webpackConfig.module.loaders.push({
+  test    : /\.scss$/,
+  exclude : null,
+  loaders : [
+    'style',
+    BASE_CSS_LOADER,
+    'postcss',
```
Comment (chris-bryant) [inline config/webpack.config.js:149]: Is PostCSS a thing we want to include at this stage?

Comment (jschroed91) [reply config/webpack.config.js:149]: Currently the main thing it's used for is autoprefixing css for cross-browser support (https://github.com/postcss/autoprefixer).

It's also handling the minification of the CSS here - I think it does a better job of it than the css-loader does.

And to be clear - this isn't doing any CSS Modules stuff - I think it's most definitely worth it to have the autoprefixer here, and it also leaves us open for including additional postcss addons in the future

File: config/webpack.config.js [hunk 2]
```diff
@@ -125,6 +130,11 @@ webpackConfig.module.loaders = [{
 
 // ------------------------------------
 // Style Loaders
+//
+// style    - This loader injects <style> tags.
+// css      - Resolves imports and returns css.
+// postcss  - Applies PostCSS processing.
+// sass     - Loads and compiles sass files.
```
Comment (jschroed91) [inline config/webpack.config.js:137]: @chris-bryant I really like what you did here w/ this comment 👍  👍  This is very helpful

**Comment by jschroed91** [comment]
@chris-bryant Thank you so much for helping out with this documentation!

Calling all @caxy/caxy-front-end-developers to review this PR! Now that we have documentation in, I think it's ready for official review.

Reminder of process - needs 2+ approvals in order to be merged. If it doesn't get that before next dev meeting, it will be put on the floor at that time.

File: config/webpack.config.js (hunk 2 above)
Comment (chris-bryant) [reply config/webpack.config.js:137]: :+1: It's probably worth noting too (and this is a bit of an assumption, but I think it's correct) that loaders are executed in reverse order when declared in an array like this. So sass -> postcss -> css -> style.

I actually didn't realize that before.

Comment (jschroed91) [reply config/webpack.config.js:137]: Yeah - I like the way the docs put it - that webpack reads right-to-left

> When chaining loaders, they are applied right to left (from the file, back).
>
> https://webpack.github.io/docs/loaders.html

**Pull request merged**